add nuker

LordMZTE opened the pull request:
closes #287

LordMZTE committed:
add nuker

5HT2 commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,132 @@
+import net.minecraft.util.math.Vec3d
+import net.minecraft.world.RaycastContext
+
+@Module.Info(name = "Nuker", category = Module.Category.PLAYER, description = "Destroys Blocks around the Player")
```
these should be split by line, like so
```kotlin
@Module.Info(
    name = "Nuker", 
    category = Module.Category.PLAYER, 
    description = "Destroys Blocks around the Player"
)
```

5HT2 commented on src/main/kotlin/me/zeroeightsix/kami/util/VectorMath.kt:
```diff
@@ -30,6 +30,9 @@ val Vec3i.asVec3d
     get() = Vec3d(this.x.toDouble(), this.y.toDouble(), this.z.toDouble())
 
+// unfortunately there are no static extension functions for java classes at this point
+fun single_vec(v: Double) = Vec3d(v, v, v)
```
This should be camelCase

LordMZTE replied:
Oops, habits from rust

JerryWester commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,132 @@
+    private fun getBoxCorner(playerPos: Vec3d, negative: Boolean = false) =
+        playerPos.add(Vec3d(hitRange, hitRange, hitRange).run { if (negative) negate() else this })
+
+    private fun startMine(blockPos: BlockPos) {
```
Clearly a bitcoin miner, smh

fee1-dead commented on src/main/kotlin/me/zeroeightsix/kami/util/VectorMath.kt:
```diff
@@ -30,6 +30,9 @@ val Vec3i.asVec3d
 val BlockPos.asVec
     get() = Vec3d(this.x.toDouble(), this.y.toDouble(), this.z.toDouble())
```
Looks like this is never used, maybe delete this?

zeroeightysix replied:
This wasn't added by @LordMZTE. It is used in Scaffold.

fee1-dead replied:
oh ok

fee1-dead commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,132 @@
+        instantMineBlocks(it.player, it.world)
+
+        if (currentBlock == null) {
+            nextBlock(it.player, it.world)
```
The function name looks like it should return a value, can you make it so that

  - Its name does not suggest a return value (findNextBlock)
  - make it return a value, so you would not access the field from the listener directly.

zeroeightysix replied:
To add to this, it is cleaner to let kotlin infer `currentBlock` being nonnull. Let `nextBlock` return the next block instead of assigning in the function, and then assign it to `currentBlock` like so:
```kotlin
currentBlock = nextBlock(it.player, it.world) ?: return@Listener
```
kotlin _should_ be able to tell that `currentBlock` is never null due to the return statement, so the `!= null` check in the next if statement isn't needed. (smart casting)

fee1-dead replied:
you will need to declare it as a local variable in this case, because kotlin thinks that another thread can be modifying this field. You would use a local variable and have the `nextBlock` function modify the field in this case.

fee1-dead commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,132 @@
+        }
+
+        if (!it.player.isCreative && currentBlock != null) {
+            val state = it.world.getBlockState(currentBlock!!)
```
If you use a local variable you can avoid the `!!` here.

fee1-dead commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,132 @@
+            val state = it.world.getBlockState(currentBlock!!)
+
+            if (selectTool) {
+                AutoTool.equipBestTool(state)
```
Braces are not needed.

fee1-dead commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,132 @@
+                Direction.DOWN
+            )
+        )
+    }
```
These two functions are repetitive, maybe you can merge these two with an extra argument?

fee1-dead reviewed (changes requested):
Good start, but I think you can still improve it, (you don't have to implement my suggested changes), it's just a suggestion

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,132 @@
+
+@Module.Info(name = "Nuker", category = Module.Category.PLAYER, description = "Destroys Blocks around the Player")
+object Nuker : Module() {
+    @Setting(name = "Hit Range")
```
This setting name is the same as the one that will be generated by fiber if you did not specify a setting name. Thus, there is no need to specify it - just `@Setting` will suffice.

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,132 @@
+    @Setting(name = "Hit Range")
+    private var hitRange: @Setting.Constrain.Range(min = 0.0, max = 8.0, step = 1.0) Double = 4.0
+
+    @Setting(name = "Through Blocks")
```
Same thing with naming here.

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,132 @@
+    @Setting(name = "Through Blocks")
+    private var throughBlocks = false
+
+    @Setting(name = "Select Tool", comment = "select the best tool for the block being broken")
```
Same thing with naming here.

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,132 @@
+    })
+
+    private var progress = 0.0
+    private var currentBlock: BlockPos? = null
```
Can you move these fields to be above all functions?

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    private fun nextBlock(player: ClientPlayerEntity, world: ClientWorld) {
+        for (block in BlockPos.method_29715(Box(getBoxCorner(player.pos), getBoxCorner(player.pos, true)))) {
```
Considering `method_29715` returns a stream, consider also using the stream API to find an entry that satisfies the `validate` function?

Something like `BlockPos.method_29715(Box(getBoxCorner(player.pos), getBoxCorner(player.pos, true))).filter { validate(player, world, it) }.findAny()`.
This returns an `Optional`, so you might want to use `Utils.kt`'s `kotlin` extension property on Optional for a nice way to make this more kotlin-y

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    private fun instantMineBlocks(player: ClientPlayerEntity, world: ClientWorld) {
+        for (block in BlockPos.method_29715(Box(getBoxCorner(player.pos), getBoxCorner(player.pos, true)))) {
```
Same thing with the Stream api here.

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,132 @@
+                hitRange
+            ) &&
+            // this stupid way of checking for unbreakable blocks was not my idea
+            (player.isCreative || !(block is CommandBlock || block is StructureBlock || block is JigsawBlock))
```
Looks like unbreakable blocks have a hardness of `-1.0f` (check `Blocks.java` (minecraft class)'s definitions.) Try checking the hardness instead of this?

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,132 @@
+            // this stupid way of checking for unbreakable blocks was not my idea
+            (player.isCreative || !(block is CommandBlock || block is StructureBlock || block is JigsawBlock))
+
+        return blockCheck && throughBlockCheck
```
Raycasting is somewhat expensive and this is a short-circuit check. If `blockCheck` is false, there is no need to calculate `throughBlockCheck` - consider first doing the block check, and, if it's false, return `false` before computing the raycast.

LordMZTE committed:
address requested changes
 rename incorrectly named function
 split module info onto multiple lines
 remove redundant setting names
 move vars before functions
 nextBlock returns a block instead of setting currentBlock
 use local variable after null check instead of !!
 remove redundant braces
 replace startMine and stopMine functions with mine function
 replace for loops with streams
 dont raycast to block if other validation fails

LordMZTE committed:
use hardness to check for unbreakable blocks

fee1-dead commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,128 @@
+    private val updateListener = Listener<TickEvent.InGame>({
+        instantMineBlocks(it.player, it.world)
+
+        if (currentBlock == null)
+            currentBlock = nextBlock(it.player, it.world)
```
```suggestion
        val currentBlock = targetBlock(it.player, it.world) ?: return@Listener
```

fee1-dead commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,128 @@
+        )
+    }
+
+    private fun nextBlock(player: ClientPlayerEntity, world: ClientWorld) =
+        BlockPos.method_29715(Box(getBoxCorner(player.pos), getBoxCorner(player.pos, true)))
+            .filter { validate(player, world, it) }.findFirst().kotlin
```
```suggestion
    private fun targetBlock(player: ClientPlayerEntity, world: ClientWorld) : BlockPos {
        val currentBlock = this.currentBlock
        if (currentBlock == null) {
            val block = BlockPos.method_29715(Box(getBoxCorner(player.pos), getBoxCorner(player.pos, true)))
            .filter { validate(player, world, it) }.findFirst().kotlin
            this.currentBlock = block
            block
        } else currentBlock
    }
```

LordMZTE replied:
why should i rename it to targetBlock?

LordMZTE replied:
resolved on discord

LordMZTE committed:
replace null check with ?: operator

LordMZTE committed:
fix instantly minable blocks as getting mistaken as unbreakable

LordMZTE committed:
add onlyInstant setting to nuker

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,128 @@
+    @Setting
+    private var throughBlocks = false
+
+    @Setting(comment = "select the best tool for the block being broken")
+    private var selectTool = false
+
+    @Setting(comment = "only break blocks that can be instantly broken")
+    private var onlyInstant = false
```
Can you start these comments with a capital letter?

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    private fun nextBlock(player: ClientPlayerEntity, world: ClientWorld) =
+        BlockPos.method_29715(Box(getBoxCorner(player.pos), getBoxCorner(player.pos, true)))
```
`BlockPos.method_29715(Box(getBoxCorner(player.pos), getBoxCorner(player.pos, true)))` appears twice - consider making a method that returns this value. Aside from removing the repetition, it'd also serve to clear up what `BlockPos.method_29715` does.

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,128 @@
+    private fun validate(player: ClientPlayerEntity, world: ClientWorld, block: BlockPos): Boolean {
+        val state = world.getBlockState(block)
+
+        val throughBlockCheck = {
+            throughBlocks ||
+                world.raycast(
+                    RaycastContext(
+                        Vec3d(player.pos.x, player.eyeY, player.pos.z),
+                        block.asVec3d.add(singleVec(0.5)),
+                        RaycastContext.ShapeType.COLLIDER,
+                        RaycastContext.FluidHandling.NONE,
+                        player
+                    )
+                ).blockPos == block
+        }
+
+        val blockCheck = {
+            !state.isAir &&
+                state.block !is FluidBlock &&
+                block.isWithinDistance(
+                    player.pos,
+                    hitRange
+                ) &&
+                (player.isCreative || state.getHardness(world, block) >= 0)
+        }
+
+        return blockCheck() && throughBlockCheck()
```
Err, sure, but that's not very pretty..

Just add a `if (!throughBlockCheck) return false` in between `throughBlocksCheck` and `blockCheck` instead of this. (Which would also simplify your last return to `return blockCheck`, as `throughBlocksCheck` is always true by the time you'd reach the last return.)

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,128 @@
+
+            if (progress == 0.0)
+                mine(block, true)
+
+
+            progress += state.calcBlockBreakingDelta(it.player, it.world, block)
```
Style: remove one blank line.

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,128 @@
+
+
+        if (!it.player.isCreative && !onlyInstant) {
+            val block = currentBlock ?: return@Listener
```
I'd rather have this return moved to where `currentBlock` gets its value.

Meaning, simplify
```kotlin
if (currentBlock == null)
    currentBlock = nextBlock(it.player, it.world)
```

to

```kotlin
val block = nextBlock(it.player, it.world).also { currentBlock = it } ?: return@Listener
```

LordMZTE replied:
i made a mistake where i wasnt validating the block before trying to break it, so by fixing this, this shortcut wouldnt work

LordMZTE committed:
implement requested nuker changes

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,126 @@
+    private val updateListener = Listener<TickEvent.InGame>({
+        instantMineBlocks(it.player, it.world)
+
+        if (currentBlock == null || validate(it.player, it.world, currentBlock!!))
```
Shouldn't this be `!validate(...)`? (If the current block is **in**valid, find the next block)

LordMZTE replied:
oh it should be, my bad. for some reason it was working tho

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -0,0 +1,126 @@
+        BlockPos.method_29715(Box(getBoxCorner(player.pos), getBoxCorner(player.pos, true)))
+            .filter { validate(player, world, it) }
+
+    /**
+     * mines a block using packets
+     * @param op if true, mining is started, if false it is stopped
+     */
```
```suggestion
    /**
     * Mines a block using packets
     * 
     * @param op if `true`, mining is started. Otherwise, it is stopped.
     */
```

LordMZTE committed:
fix incorrect check
improve grammar on doc comment

LordMZTE committed:
add option to nuker to use player reach

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -38,6 +43,15 @@ object Nuker : Module() {
 
+    private val range
+        get() =
+            // this doesn't look that great, but i can't think of a cleaner way to do this
```
This doesn't look bad - it's fine.

Do remove the comment though. Next time just add a comment on your PR instead of in the code!

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -120,7 +134,7 @@ object Nuker : Module() {
             block.isWithinDistance(
                 player.pos,
-                hitRange
+                range
```
`validate` is called quite often, so we don't want to recalculate `range` every time. Add a `range` parameter to `validate`, and let `validate`'s caller compute the range once. Preferably add the parameter to the callers as well, and only calculate `range` in the event listener.

LordMZTE committed:
nuker will only calculate player range once a tick

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -45,15 +45,19 @@ object Nuker : Module() {
             if (usePlayerRange) mc.interactionManager?.reachDistance?.toDouble() ?: hitRange
             else hitRange
 
+    // This exists to use the range without having to recalculate `range` a lot.
+    // Instead it is calculated once at the start of the tick event
+    private var currentRange = range
+
```
Oh god, no.
Please just use a local variable and pass it on through parameters.

LordMZTE replied:
alright then, i'll change that. i'm trying my best

LordMZTE committed:
use parameter instead of field for current range in nuker

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/module/Nuker.kt:
```diff
@@ -105,19 +101,19 @@ object Nuker : Module() {
 
-    private fun nextBlock(player: ClientPlayerEntity, world: ClientWorld) =
-        getValidBlocks(player, world).findFirst().kotlin
+    private fun nextBlock(range: Double, player: ClientPlayerEntity, world: ClientWorld) =
```
This is a nitpick, but please move the range to be the last parameter in the list (and do the same for the other methods taking a range)

If there is an parameter with a default value, insert `range` before that parameter.

This because the range is the least significant parameter of this function.

LordMZTE committed:
make range last parameter

zeroeightysix reviewed (changes requested):
Please fix build.

LordMZTE commented:
oh i forgot that this got updated to 1.16.4

LordMZTE committed:
update to 1.14

LordMZTE committed:
Merge remote-tracking branch 'original/fabric' into feature/nuker

zeroeightysix commented:
Thanks!
